virtcontainers/clh.go

```diff
@@ -424,22 +432,84 @@ func (clh *cloudHypervisor) hypervisorConfig() HypervisorConfig {
 }
 
 func (clh *cloudHypervisor) resizeMemory(reqMemMB uint32, memoryBlockSizeMB uint32, probe bool) (uint32, memoryDevice, error) {
-	clh.Logger().WithField("function", "resizeMemory").Warn("not supported")
-	return 0, memoryDevice{}, nil
+	if reqMemMB == 0 {
+		return 0, memoryDevice{}, errors.New("Can not resize memory to 0")
+	}
+
+	info, err := clh.vmInfo()
+	if err != nil {
+		return 0, memoryDevice{}, err
+	}
+
+	currentMem := utils.MemUnit(info.Config.Memory.Size) * utils.Byte
+	newMem := utils.MemUnit(reqMemMB) * utils.MiB
+
+	// Early check to verify if boot memory is the same as requested
+	if currentMem == newMem {
+		clh.Logger().WithField("memory", reqMemMB).Debugf("VM already has requested memory")
+		return 0, memoryDevice{}, nil
+	}
+
+	if currentMem > newMem {
+		clh.Logger().Warn("Remove memory is not supported, nothing to do")
+		return 0, memoryDevice{}, nil
+	}
+
+	blockSize := utils.MemUnit(memoryBlockSizeMB) * utils.MiB
+	hotplugSize := (newMem - currentMem).AlignMem(blockSize)
+
+	// Update memory request to increase memory aligned block
+	alignedRequest := currentMem + hotplugSize
+	if newMem != alignedRequest {
+		clh.Logger().WithFields(log.Fields{"request": newMem, "aligned_request": alignedRequest}).Debug("aligning VM memory request")
+		newMem = alignedRequest
+	}
+
+	// Check if memory is the same as requested, a second check is done
+	// to consider the memory request now that is updated to be memory aligned
+	if currentMem == newMem {
+		clh.Logger().WithFields(log.Fields{"current_memory": currentMem, "new_memory": newMem}).Debug("VM already has requested memory(after alignment)")
+		return 0, memoryDevice{}, nil
+	}
+
+	cl := clh.client()
+	ctx, cancelResize := context.WithTimeout(context.Background(), clhAPITimeout*time.Second)
+	defer cancelResize()
+
+	// OpenApi does not support uint64, convert to int64
+	resize := chclient.VmResize{DesiredRam: int64(newMem.ToBytes())}
+	clh.Logger().WithFields(log.Fields{"current_memory": currentMem, "new_memory": newMem}).Debug("updating VM memory")
+	if _, err = cl.VmResizePut(ctx, resize); err != nil {
+		clh.Logger().WithFields(log.Fields{"current_memory": currentMem, "new_memory": newMem}).Warnf("failed to update memory %s", openAPIClientError(err))
+		err = fmt.Errorf("Failed to resize memory from %d to %d: %s", currentMem, newMem, openAPIClientError(err))
+		return 0, memoryDevice{}, openAPIClientError(err)
+	}
+
+	info, err = clh.vmInfo()
+	if err != nil {
+		clh.Logger().WithError(err).Warn("vmInfo failed")
+		return 0, memoryDevice{}, openAPIClientError(err)
+	}
+
+	newMem = utils.MemUnit(info.Config.Memory.Size) * utils.Byte
+	addedMemory := newMem - currentMem
+
+	if hotplugSize != addedMemory {
+		errors.Errorf("Added memory different than expected %d != %d", newMem, addedMemory)
+	}
```

Comment:
Hmm, so we have to make _two_ calls? If `VmResizePut()` could in future return the actual value set, that would save an HTTP request I think?

/cc @sboeuf.

Reply:
No there's no need for that because if the call didn't return any error, that means the expected size is now the current one. So basically I don't think there's a need for calling into `VmInfo` to verify the size.

Reply:
Thanks @sboeuf - so you can zap that bit @jcvenegas ;)